File: infra/tools/db/audit_database.js
/**
 * Script para auditar o banco de dados Supabase e listar todas as tabelas, views e RPCs
 */

const { Client } = require('pg');
require('dotenv').config({ path: require('path').join(__dirname, '../../.env') });

const dbConfig = {
  host: process.env.GF_DB_HOST || 'db.vmoxzesvjcfmrebagcwo.supabase.co',
  port: process.env.GF_DB_PORT || 5432,
  user: process.env.GF_DB_USER || 'postgres',
  password: process.env.GF_DB_PASSWORD || 'Guigui1309@',
  database: process.env.GF_DB_NAME || 'postgres',
  ssl: { rejectUnauthorized: false }
};

async function auditDatabase() {
  const client = new Client(dbConfig);
  
  try {
    await client.connect();
    console.log('✅ Conectado ao banco de dados\n');

    // Listar tabelas
    const tablesResult = await client.query(`
      SELECT table_name 
      FROM information_schema.tables
      WHERE table_schema = 'public'
      ORDER BY table_name;
    `);
    const tables = tablesResult.rows.map(r => r.table_name);

    // Listar views
    const viewsResult = await client.query(`
      SELECT table_name 
      FROM information_schema.views
      WHERE table_schema = 'public'
      ORDER BY table_name;
    `);
    const views = viewsResult.rows.map(r => r.table_name);

    // Listar RPCs
    const rpcsResult = await client.query(`
      SELECT routine_name 
      FROM information_schema.routines
      WHERE routine_schema = 'public'
      AND routine_type = 'FUNCTION'
      ORDER BY routine_name;
    `);
    const rpcs = rpcsResult.rows.map(r => r.routine_name);

    console.log('=== TABELAS ===');
    tables.forEach(t => console.log(`  - ${t}`));

    console.log(`\n=== VIEWS (${views.length}) ===`);
    views.forEach(v => console.log(`  - ${v}`));

    console.log(`\n=== RPCs/FUNÇÕES (${rpcs.length}) ===`);
    rpcs.forEach(r => console.log(`  - ${r}`));

    // Verificar tabelas gf_
    const gfTables = tables.filter(t => t.startsWith('gf_'));
    console.log(`\n=== TABELAS gf_ (${gfTables.length}) ===`);
    gfTables.forEach(t => console.log(`  - ${t}`));

    // Verificar views esperadas
    const expectedViews = [
      'v_dashboard_kpis',
      'v_driver_last_position',
      'v_active_trips',
      'v_route_costs',
      'v_driver_last_status'
    ];
    console.log(`\n=== VIEWS ESPERADAS ===`);
    expectedViews.forEach(v => {
      const exists = views.includes(v);
      console.log(`  ${exists ? '✅' : '❌'} ${v}`);
    });

    // Verificar RPCs esperadas
    const expectedRpcs = [
      'rpc_generate_route_stops',
      'rpc_optimize_route_google',
      'rpc_validate_boarding',
      'gf_map_snapshot_full'
    ];
    console.log(`\n=== RPCs ESPERADAS ===`);
    expectedRpcs.forEach(r => {
      const exists = rpcs.includes(r);
      console.log(`  ${exists ? '✅' : '❌'} ${r}`);
    });

    // Tabelas esperadas gf_
    const expectedGfTables = [
      'gf_notifications',
      'gf_assistance_requests',
      'gf_gamification_scores',
      'gf_vehicle_costs',
      'gf_boarding_tokens',
      'gf_boarding_events',
      'gf_employee_company'
    ];
    console.log(`\n=== TABELAS gf_ ESPERADAS ===`);
    expectedGfTables.forEach(t => {
      const exists = tables.includes(t);
      console.log(`  ${exists ? '✅' : '❌'} ${t}`);
    });

    // JSON de auditoria
    const audit = {
      web: {
        stack: "next14",
        tem_12_abas: true,
        tem_mapa_google: true,
        usa_app_router: true
      },
      mobile: {
        stack: "flutter3.24",
        features: ["driver", "passenger", "operator", "admin"],
        tem_tracking_service: false // verificar depois
      },
      db: {
        schema: "v7.4",
        tabelas: tables,
        views: views,
        rpcs: rpcs,
        faltando: {
          views: expectedViews.filter(v => !views.includes(v)),
          rpcs: expectedRpcs.filter(r => !rpcs.includes(r)),
          tabelas_gf: expectedGfTables.filter(t => !tables.includes(t))
        }
      }
    };

    console.log(`\n=== JSON AUDITORIA ===`);
    console.log(JSON.stringify(audit, null, 2));

  } catch (error) {
    console.error('❌ Erro:', error.message);
    process.exit(1);
  } finally {
    await client.end();
  }
}

auditDatabase();

File: infra/tools/db/audit_database.py
#!/usr/bin/env python3
"""
Script para auditar o banco de dados Supabase e listar todas as tabelas, views e RPCs
"""

import os
import sys
import psycopg2
from psycopg2.extras import RealDictCursor

def get_db_connection():
    """Conecta ao banco usando variáveis de ambiente"""
    host = os.getenv('GF_DB_HOST', 'db.vmoxzesvjcfmrebagcwo.supabase.co')
    port = os.getenv('GF_DB_PORT', '5432')
    user = os.getenv('GF_DB_USER', 'postgres')
    password = os.getenv('GF_DB_PASSWORD', 'Guigui1309@')
    dbname = os.getenv('GF_DB_NAME', 'postgres')
    
    try:
        conn = psycopg2.connect(
            host=host,
            port=port,
            user=user,
            password=password,
            dbname=dbname,
            sslmode='require'
        )
        return conn
    except Exception as e:
        print(f"Erro ao conectar: {e}", file=sys.stderr)
        sys.exit(1)

def query_tables(conn):
    """Lista todas as tabelas"""
    with conn.cursor(cursor_factory=RealDictCursor) as cur:
        cur.execute("""
            SELECT table_name 
            FROM information_schema.tables
            WHERE table_schema = 'public'
            ORDER BY table_name;
        """)
        return [row['table_name'] for row in cur.fetchall()]

def query_views(conn):
    """Lista todas as views"""
    with conn.cursor(cursor_factory=RealDictCursor) as cur:
        cur.execute("""
            SELECT table_name 
            FROM information_schema.views
            WHERE table_schema = 'public'
            ORDER BY table_name;
        """)
        return [row['table_name'] for row in cur.fetchall()]

def query_rpcs(conn):
    """Lista todas as RPCs (funções)"""
    with conn.cursor(cursor_factory=RealDictCursor) as cur:
        cur.execute("""
            SELECT routine_name 
            FROM information_schema.routines
            WHERE routine_schema = 'public'
            AND routine_type = 'FUNCTION'
            ORDER BY routine_name;
        """)
        return [row['routine_name'] for row in cur.fetchall()]

def main():
    conn = get_db_connection()
    
    try:
        tables = query_tables(conn)
        views = query_views(conn)
        rpcs = query_rpcs(conn)
        
        print("=== TABELAS ===")
        for table in tables:
            print(f"  - {table}")
        
        print(f"\n=== VIEWS ===")
        for view in views:
            print(f"  - {view}")
        
        print(f"\n=== RPCs (FUNÇÕES) ===")
        for rpc in rpcs:
            print(f"  - {rpc}")
        
        # Verificar tabelas específicas gf_
        gf_tables = [t for t in tables if t.startswith('gf_')]
        print(f"\n=== TABELAS gf_ ({len(gf_tables)}) ===")
        for table in gf_tables:
            print(f"  - {table}")
        
        # Verificar views esperadas
        expected_views = [
            'v_dashboard_kpis',
            'v_driver_last_position',
            'v_active_trips',
            'v_route_costs',
            'v_driver_last_status'
        ]
        print(f"\n=== VIEWS ESPERADAS ===")
        for view in expected_views:
            status = "✅" if view in views else "❌"
            print(f"  {status} {view}")
        
        # Verificar RPCs esperadas
        expected_rpcs = [
            'rpc_generate_route_stops',
            'rpc_optimize_route_google',
            'rpc_validate_boarding',
            'gf_map_snapshot_full'
        ]
        print(f"\n=== RPCs ESPERADAS ===")
        for rpc in expected_rpcs:
            status = "✅" if rpc in rpcs else "❌"
            print(f"  {status} {rpc}")
        
        # Tabelas esperadas gf_
        expected_gf_tables = [
            'gf_notifications',
            'gf_assistance_requests',
            'gf_gamification_scores',
            'gf_vehicle_costs',
            'gf_boarding_tokens',
            'gf_boarding_events',
            'gf_employee_company'
        ]
        print(f"\n=== TABELAS gf_ ESPERADAS ===")
        for table in expected_gf_tables:
            status = "✅" if table in tables else "❌"
            print(f"  {status} {table}")
        
    finally:
        conn.close()

if __name__ == '__main__':
    main()

